### PR TITLE
PR CI Improvments

### PR DESCRIPTION
**Pull request by brandonpage**
- Run Tests Conditionally, based on modified files
- Fix Danger for all users

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -22,6 +22,14 @@ jobs:
 
       - checkout
 
+      - run:
+          name: Determine Tests to Run
+          shell: /bin/bash
+          command:  |
+            LIBS_TO_TEST=$(ruby .circleci/gitChangedLibs.rb)
+            echo "export LIBS_TO_TEST=${LIBS_TO_TEST}" >> "${BASH_ENV}"
+            echo -e "\n\nLibraries to Test-> ${LIBS_TO_TEST}"
```
Comment (brandonpage) [inline .circleci/config.yml]: Print libraries to be tested to quick reference while CI is running.

File: .circleci/config.yml [hunk 2]
```diff
@@ -40,48 +48,88 @@ jobs:
...
           name: Combine lib Test Restuls
           command: junit-merge libs/**/build/outputs/androidTest-results/connected/*.xml
           when: always
       - run:
           name:  Run Danger
-          command: danger --dangerfile=.circleci/Dangerfile --danger_id=ci/circleci
           when: always
+          command: |
+            if [ -n "$CIRCLE_PR_NUMBER" ]; then
+              DANGER_GITHUB_API_TOKEN="c21349d8a97e1bf9cdd9""301fd949a83db862216b" danger --dangerfile=.circleci/Dangerfile --danger_id=ci/circleci
```
Comment (brandonpage) [inline .circleci/config.yml]: Unfortunately, this appears to be the only workaround to get danger to run consistently in CircleCI.  The token here only has public_repo access so if someone takes it I don't foresee them being able to do anything nefarious.

File: .circleci/gitChangedLibs.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  path = prfile["filename"]
+  for lib in $libs
+    if path.include? lib
+      libs = libs.add(lib)
+    end
+  end
+end
+
+libs_ar = libs.to_a()
+if !libs_ar.empty?() && libs_ar.include?("SalesforceSDK")
```
Comment (brandonpage) [inline .circleci/gitChangedLibs.rb]: If files are chanced in Core, run all tests.

File: .circleci/gitChangedLibs.rb [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+  libs_ar = $libs.to_a()
+elsif libs_ar.include?("SmartStore") && !libs_ar.include?("SmartSync")
```
Comment (brandonpage) [inline .circleci/gitChangedLibs.rb]: If changes in SmartStore, test SmartSync too.

File: .circleci/gitChangedLibs.rb [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+  libs_ar.push("SmartSync")
+end
```
Comment (brandonpage) [inline .circleci/gitChangedLibs.rb]: Any other dependencies I missed?

**Comment by brandonpage** [comment]
If none of the libraries are changed should we run all tests?  Or none?  

Right now I have it running none because the only scenarios I could think of are: CI updates, Readme update, test app changes, configuration updates.

**Comment by wmathurin** [comment]
SalesforceHybrid and SalesforceReact both depend on SmartSync.

**Comment by brandonpage** [comment]
## Danger has errored
[!] Invalid `Dangerfile` file: No JUnit file was found at merged-test-results.xml. Updating the Danger gem might fix the issue. Your Danger version: 5.5.3, latest Danger version: 5.5.4

```
 #  from .circleci/Dangerfile:1
 #  -------------------------------------------
 >  junit.parse "merged-test-results.xml"
 #  junit.show_skipped_tests = true
 #  -------------------------------------------
```

<p align="right" data-meta="generated_by_ci/circleci">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: .circleci/gitChangedLibs.rb (hunk 3 above)
Comment (wmathurin) [reply .circleci/gitChangedLibs.rb]: SalesforceSDK depends on SalesforceAnalytics (see https://github.com/forcedotcom/SalesforceMobileSDK-Android/blob/master/libs/SalesforceSDK/build.gradle#L17).
So line 22, we should test for SalesforceAnalytics.
We should have another elsif for SalesforceSDKCore that would push SmartStore, SmartSync, SalesforceHybridSDK, SalesforceReact.

Comment (wmathurin) [reply .circleci/gitChangedLibs.rb]: We could simplify the code:

```ruby
libsTopoSorted = ["SalesforceAnalytics", "SalesforceSDK", "SmartStore", "SmartSync", "SalesforceHybridSDK", "SalesforceReact"] 
libsModified = libs.to_a()
libsToTest = libsTopoSorted.slice(libsModified.map{|l| libsTopoSorted.find_index(l)}.min() .. libsTopoSorted.length)
```
libsTopoSorted is topologically sorted - x before y means x does not depend on y.
It looks for the modified library that has the smallest index in libsTopoSorted.
Then tests that library and all the ones that follow in libsTopoSorted.

**Pull request merged**